storage/etcd/etcd.go

```diff
@@ -0,0 +1,472 @@
+	res, err := c.db.Get(context.TODO(), connectorPrefix, clientv3.WithPrefix())
+	if err != nil {
+		return nil, err
+	}
+	for _, v := range res.Kvs {
+		var c storage.Connector
+		if err = json.Unmarshal(v.Value, &c); err != nil {
+			return nil, err
+		}
+		connectors = append(connectors, c)
+	}
+	return connectors, nil
+}
+
+func (c *conn) GetKeys() (keys storage.Keys, err error) {
+	res, err := c.db.Get(context.TODO(), keysName)
+	if err != nil {
+		return keys, err
+	}
+	if res.Count > 0 && len(res.Kvs) > 0 {
+		err = json.Unmarshal(res.Kvs[0].Value, &keys)
+	}
+	return keys, err
+}
+
+func (c *conn) UpdateKeys(updater func(old storage.Keys) (storage.Keys, error)) error {
+	key := keysName
+	ctx := context.TODO()
+	return c.txnUpdate(ctx, key, func(currentValue []byte) ([]byte, error) {
+		var current storage.Keys
+		if len(currentValue) > 0 {
+			if err := json.Unmarshal(currentValue, &current); err != nil {
+				return nil, err
+			}
+		}
+		updated, err := updater(current)
+		if err != nil {
+			return nil, err
+		}
+		return json.Marshal(updated)
+	})
+}
+
+func (c *conn) deleteKey(prefix string, id string) error {
+	key := prefix + id
+	res, err := c.db.Delete(context.TODO(), key)
+	if err != nil {
+		return err
+	}
+	if res.Deleted == 0 {
+		return storage.ErrNotFound
+	}
+	return nil
+}
+
+func offlineSessionName(userID string, connID string) string {
+	return strings.ToLower(userID + "|" + connID)
+}
+
+func (c *conn) getOne(key string, value interface{}) error {
+	r, err := c.db.Get(context.TODO(), key)
+	if err != nil {
+		return err
+	}
+	if r.Count == 0 {
+		return storage.ErrNotFound
+	}
+	return json.Unmarshal(r.Kvs[0].Value, value)
+}
+
+func (c *conn) listAuthRequests() (reqs []AuthRequest, err error) {
+	res, err := c.db.Get(context.TODO(), authRequestPrefix, clientv3.WithPrefix())
+	if err != nil {
+		return reqs, err
+	}
+	for _, v := range res.Kvs {
+		var r AuthRequest
+		if err = json.Unmarshal(v.Value, &r); err != nil {
+			return reqs, err
+		}
+		reqs = append(reqs, r)
+	}
+	return reqs, nil
+}
+
+func (c *conn) listAuthCodes() (codes []AuthCode, err error) {
+	res, err := c.db.Get(context.TODO(), authCodePrefix, clientv3.WithPrefix())
+	if err != nil {
+		return codes, err
+	}
+	for _, v := range res.Kvs {
+		var c AuthCode
+		if err = json.Unmarshal(v.Value, &c); err != nil {
+			return codes, err
+		}
+		codes = append(codes, c)
+	}
+	return codes, nil
+}
+
+func (c *conn) txnCreate(ctx context.Context, key string, value interface{}) error {
+	b, err := json.Marshal(value)
+	if err != nil {
+		return err
+	}
+	txn := c.db.Txn(ctx)
+	res, err := txn.
+		If(clientv3.Compare(clientv3.CreateRevision(key), "=", 0)).
+		Then(clientv3.OpPut(key, string(b))).
+		Commit()
+	if err != nil {
+		return err
+	}
+	if !res.Succeeded {
+		return storage.ErrAlreadyExists
+	}
+	return nil
+}
+
+func (c *conn) txnUpdate(ctx context.Context, key string, update func(current []byte) ([]byte, error)) error {
```

Comment:
cc @xiang90 mind taking a look at this?

Reply:
looks right to me. 

we could do auto retry inside the txnUpdate by calling update func multiple times when there are conflicts (maybe it is already wrapped?).

Reply:
I dont think auto-retry should be applied here. At least thats how the current test is structured ( also how sql transaction works ).
Although IIUC there are differences on the order of transaction between sql and non-sql storages now ( k8s or etcd ). In sql the first started transaction win vs the first finished transaction win in others.

Reply:
@dqminh understood. thanks.